### DISCUSSION TITLE
Separating the DAG parsing process from the scheduler

### DISCUSSION THREAD

**Discussion by vbambuch** [apache/airflow#23721] [General] [answered]
Hello Airflow community,

from what I learned from the new Airflow concept of scheduling using serialized DAGs the scheduler now does the following actions:
* regularly parses all DAGs and stores their serialized version into the DB
* regularly schedules new tasks based on the serialized data in the DB
* those two actions are not strictly dependent on each other (but of course without the serialized DAGs the scheduler cannot schedule) 

**The problem**:
The DAG parsing is the heaviest thing the scheduler does and therefore it requires a lot of resources (CPU). If I run multiple schedulers they all do the same: parsing DAGs and scheduling new tasks. If I run multiple schedulers they all parse the same amount of DAGs but they don't schedule the same DAGs (obviously).

**The idea**:
So I thought about moving (by tweaking the `min_file_process_interval` and `scheduler_heartbeat_sec` attributes) the parsing responsibility to dedicated instance(s) and the rest would handle the scheduling only. The parsing would be still the same with increasing demand for resources while adding more DAGs but the demand would be more or less linear. On the other hand, the scheduling is highly dynamic based on which DAGs are supposed to run so there is the opportunity to utilize the multiple schedulers feature.
By doing this I'm able to decrease the number of resources needed to run multiple schedulers.

**More conservative way**:
Another option is to decrease the `min_file_process_interval` because the parsing is done in multiple instances and keep the high resources.

What is your opinion on **The idea**? Do you see any issues in this architecture?
I just don't see a point in running the scheduler on the high scale if most of the work is redundant. I have already made a small POC and it works nicely.

**Comment by potiuk** [comment, marked as answer]
There is a new feature in Airlfow 2.3.0 (partially implemented https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-43+DAG+Processor+separation which implements just this. 

You can use it even now after you migrate to 2.3.0:  https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#standalone-dag-processor

There is also a work in progress (https://github.com/apache/airflow/pull/23711) to add it as an  option to the official Helm Chart.

**Reply by vbambuch** [reply]
Nice, thank you. I'll take a look at it. 👍
